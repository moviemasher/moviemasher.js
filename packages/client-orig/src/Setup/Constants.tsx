import React from 'react'
import { UnknownRecord } from "@moviemasher/lib-core"

import { JsxElement } from "../Framework/Framework"


function ConstEmptyElementFunction(props: UnknownRecord) { return null }

export const EmptyElement: JsxElement = <ConstEmptyElementFunction key="empty-element" />
