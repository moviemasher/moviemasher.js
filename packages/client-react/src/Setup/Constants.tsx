import React from 'react'
import { UnknownObject } from "@moviemasher/moviemasher.js"

import { UnknownElement, ReactResult } from "../declarations"

function ConstEmptyElementFunction(props: UnknownObject): ReactResult { return null }

export const EmptyElement: UnknownElement = <ConstEmptyElementFunction/>
