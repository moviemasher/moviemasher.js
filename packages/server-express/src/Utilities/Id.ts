// import { generateUUIDv4 } from "@bitjourney/uuid-v4"
let idUniqueCounter = 0
export const idUnique = () => {
  return `idUnique-${idUniqueCounter++}` //generateUUIDv4()
}