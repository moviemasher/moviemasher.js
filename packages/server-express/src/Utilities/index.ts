export * from "./Directory"
export * from "./Expand"
export * from "./NodeLoader"
export * from "./Rendering"
