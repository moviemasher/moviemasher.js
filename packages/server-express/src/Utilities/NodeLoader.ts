import fs from 'fs'
import https from 'https'
import http from 'http'
import path from 'path'
import md5 from 'md5'
import {
  EmptyMethod, GraphFile, GraphType, LoadedInfo, isPreloadableDefinition,
  LoadTypes, Errors, LoaderClass, GraphFileType, LoaderFile, Definition,
  LoaderSource,
  isAboveZero, isLoadType, assertPopulatedString, isUpdatableDurationDefinition, isUpdatableDimensionsDefinition
} from '@moviemasher/moviemasher.js'

import { BasenameCache, ExtensionLoadedInfo } from '../Setup/Constants'
import { probingInfoPromise } from '../Command/Probing'

export class NodeLoader extends LoaderClass {
  constructor(
    public cacheDirectory: string,
    public filePrefix: string,
    public defaultDirectory: string,
    public validDirectories: string[]
  ) {
    super()
    if (!cacheDirectory) throw Errors.invalid.url + 'cacheDirectory'
    if (!filePrefix) throw Errors.invalid.url + 'filePrefix'
    if (!defaultDirectory) throw Errors.invalid.url + 'defaultDirectory'
  }


  private applyLoadedInfo(graphFile: GraphFile, loadedInfo: LoadedInfo): void {
    const { definition } = graphFile
    const { duration, width, height } = loadedInfo
    if (isUpdatableDurationDefinition(definition) && !isAboveZero(definition.duration)) {
      if (isAboveZero(duration)) definition.duration = duration
      // console.log(this.constructor.name, "applyLoadedInfo duration", definition.id, definition.duration)
    }
    if (isUpdatableDimensionsDefinition(definition) && !isAboveZero(definition.width)) {
      if (isAboveZero(width)) definition.width = width
      if (isAboveZero(height)) definition.height = height
      // console.log(this.constructor.name, "applyLoadedInfo dimensions", definition.id, `${width}x${height}`)
    }
  }

  protected override filePromise(key: string, graphFile: GraphFile): LoaderFile {
    const { definition } = graphFile
    const definitions = new Map<string, Definition>()
    definitions.set(definition.id, definition)
    const preloaderSource: LoaderSource = { loaded: false, definitions }
    if (fs.existsSync(key)) {
      preloaderSource.promise = this.updateSources(key, preloaderSource, graphFile)
    } else {
      const { filePrefix } = this
      if (key.startsWith(filePrefix)) throw Errors.uncached + ' filePromise ' + key

      preloaderSource.promise = this.writePromise(graphFile, key).then(() => {
        return this.updateSources(key, preloaderSource, graphFile)
      })
    }
    return preloaderSource as LoaderFile
  }

  getFile(_graphFile: GraphFile): any { throw Errors.unimplemented + 'getFile' }

  graphType = GraphType.Mash

  key(graphFile: GraphFile): string {
    const { type, file } = graphFile
    assertPopulatedString(file)

    const { cacheDirectory, filePrefix, defaultDirectory, validDirectories } = this
    if (LoadTypes.map(String).includes(type)) {
      if (file.includes('://')) return path.resolve(
        cacheDirectory, md5(file), `${BasenameCache}${path.extname(file)}`
      )

      const resolved = path.resolve(filePrefix, defaultDirectory, file)
      const directories = [defaultDirectory, ...validDirectories]
      const prefixes = [path.resolve(cacheDirectory), ...directories.map(dir => path.resolve(filePrefix, dir))]
      const valid = prefixes.some(prefix => resolved.startsWith(prefix))
      if (!valid) throw Errors.invalid.url + resolved

      return resolved
    }
    if (file.startsWith('/')) console.trace(this.constructor.name, "key", file, type, md5(file))
    return path.resolve(cacheDirectory, md5(file), `${BasenameCache}.${type}`)
  }

  private remotePromise(key: string, url: string): Promise<void> {
    const promise: Promise<void> = new Promise((resolve, reject) => {
      const callback = (res: http.IncomingMessage) => {
        const stream = fs.createWriteStream(key)
        res.pipe(stream)
        stream.on('finish', () => {
          stream.close()
          resolve()
        })
        stream.on('error', (error) => { reject(error) })
      }
      if (url.startsWith('https://')) https.get(url, callback)
      else http.get(url, callback)
    })
    return promise
  }

  private updateableDefinitions(preloaderSource: LoaderSource): Definition[] {
    const definitions = [...preloaderSource.definitions.values()]
    const preloadableDefinitions = definitions.filter(definition => {
      return isPreloadableDefinition(definition)
    })

    return preloadableDefinitions.filter(definition => {
      const trimmable = isUpdatableDurationDefinition(definition)
      if (trimmable && !isAboveZero(definition.duration)) return true

      return isUpdatableDimensionsDefinition(definition) && !isAboveZero(definition.width)
    })
  }


  private updateSources(key: string, preloaderSource: LoaderSource, graphFile: GraphFile): Promise<void> {
    preloaderSource.loaded = true
    preloaderSource.definitions.forEach(definition => {
      if (!isPreloadableDefinition(definition)) return

      if (!definition.source.startsWith('http')) definition.source = key
    })

    const { type } = graphFile
    if (!isLoadType(type)) return Promise.resolve()

    const neededDefinitions = this.updateableDefinitions(preloaderSource)
    if (!neededDefinitions.length) return Promise.resolve()

    const preloaderFile = preloaderSource as LoaderFile

    const infoPath = path.join(this.cacheDirectory, `${md5(key)}.${ExtensionLoadedInfo}`)
    // console.log(this.constructor.name, "fileInfoPromise", infoPath)
    return probingInfoPromise(key, infoPath).then(info => {
      preloaderFile.loadedInfo = info
      return info
    }).then(loadedInfo => { this.applyLoadedInfo(graphFile, loadedInfo) })
  }

  private writePromise(graphFile: GraphFile, key: string): Promise<void> {
    const { file, type } = graphFile
    const loadable = LoadTypes.map(String).includes(type)
    const dirname = path.dirname(key)
    // console.log(this.constructor.name, "writePromise", dirname)
    let promise = fs.promises.mkdir(dirname, { recursive: true }).then(EmptyMethod)

    if (loadable) {
      if (file.startsWith('http')) {
        promise = promise.then(() => this.remotePromise(key, file))
      } else throw Errors.uncached + file // local files should already exist!
    } else {
      // console.log(this.constructor.name, "writePromise writeFile", key, file)
      const data = type === GraphFileType.Png ? Buffer.from(file, 'base64') : file
      promise = promise.then(() => fs.promises.writeFile(key, data))
    }
    return promise
  }
}
