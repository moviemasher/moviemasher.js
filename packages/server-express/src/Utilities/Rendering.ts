import {
  DefinitionObject, DefinitionObjects,
  MashObject, TrackObject, ClipObject,
  DefinitionType, LoadType, TrackType, ValueObject, CommandOutputs, OutputType,
  RenderingInput, RenderingCommandOutput, NumberObject, outputDefaultPopulate, VisibleClipObject, visibleClipDefault,
} from "@moviemasher/moviemasher.js"


import { idUnique } from "./Id"

export const renderingInput = (definition: DefinitionObject, clipObject: ValueObject = {}): RenderingInput => {
  const { type, id } = definition
  const definitionObject = {
    ...definition,
    type: type === DefinitionType.VideoSequence ? DefinitionType.Video : type
  }
  const definitions: DefinitionObjects = [definitionObject]
  const clip: ClipObject = renderingClipFromDefinition(definitionObject, clipObject)
  const track: TrackObject = {
    trackType: String(type) === String(DefinitionType.Audio) ? TrackType.Audio : TrackType.Video,
    clips: [clip]
  }
  const tracks: TrackObject[] = [track]
  const mashObject = { id }
  const mash: MashObject = { ...mashObject, tracks }
  return { mash, definitions }
}

export const renderingDefinitionTypeCommandOutputs = (definitionType: DefinitionType) => {
  const outputs: CommandOutputs = []

  // TODO: support waveform generation
  // TODO: support font uploading
  switch (definitionType) {
    case DefinitionType.Audio: {
      outputs.push({ outputType: OutputType.Audio })
      // outputs.push({ outputType: OutputType.Waveform })
      break
    }
    case DefinitionType.Image: {
      outputs.push({ outputType: OutputType.Image })
      break
    }
    case DefinitionType.VideoSequence: {
      outputs.push({ outputType: OutputType.Audio })
      outputs.push({ outputType: OutputType.Image })
      outputs.push({ outputType: OutputType.ImageSequence })
      // outputs.push({ outputType: OutputType.Waveform })
      break
    }
    case DefinitionType.Font: {
      // outputs.push({ outputType: OutputType.Font })
      break
    }
  }
  return outputs
}

export const renderingCommandOutputs = (commandOutputs: CommandOutputs): CommandOutputs => {
  const counts: NumberObject = {}
  return commandOutputs.map(output => {
    const { outputType } = output
    const populated = outputDefaultPopulate(output)
    if (!counts[outputType]) {
      counts[outputType] = 1
    } else {
      populated.basename ||= `${outputType}-${counts[outputType]}`
      counts[outputType]++
    }
    return populated
  })
}

export const renderingOutputFile = (commandOutput: RenderingCommandOutput, extension?: string): string => {
  const { basename, format, extension: outputExtension, outputType } = commandOutput
  const ext = extension || outputExtension || format
  return `${basename || outputType}.${ext}`
}

export const renderingSource = (commandOutput?: RenderingCommandOutput): string => {
  if (!commandOutput) {
    // console.log("renderingSource with no commandOutput")
    return ''
  }

  const { format, extension, outputType } = commandOutput
  const ext = extension || format
  if (outputType === OutputType.ImageSequence) return ''

  return `${outputType}.${ext}`
}



export const renderingInputFromRaw = (loadType: LoadType, source: string, clip: ValueObject = {}): RenderingInput => {
  const definitionId = clip.id || source 
  const definition = renderingDefinitionObject(loadType, source, String(definitionId), String(clip.label))
  return renderingInput(definition, clip)
}


export const renderingClipFromDefinition = (definition: DefinitionObject, overrides: ValueObject = {}): ClipObject => {
  const { id, type } = definition
  const { id: _, containerId: suppliedContainerId, ...rest } = overrides
  const contentId = id || type
  const supplied = suppliedContainerId ? String(suppliedContainerId) : undefined
  const containerId = type === 'audio' ? '' : supplied
  const definitionId = visibleClipDefault.id
  const visibleClipObject: VisibleClipObject = {
    definitionId, contentId, content: rest, containerId
  }
  // console.log("renderingClipFromDefinition", overrides, visibleClipObject)
  return visibleClipObject
}

export const renderingDefinitionObject = (loadType: LoadType, source: string, definitionId?: string, label?: string): DefinitionObject => {
  const type: DefinitionType = definitionTypeFromRaw(loadType)
  // console.log("renderingDefinitionObject", definitionId, typeof definitionId)
 
  const id = definitionId || idUnique()
  const definition: DefinitionObject = { id, type, source, label }
  return definition
}



export const definitionTypeFromRaw = (loadType: LoadType): DefinitionType => {
  switch (loadType) {
    case LoadType.Audio: return DefinitionType.Audio
    case LoadType.Video: return DefinitionType.VideoSequence
    case LoadType.Image: return DefinitionType.Image
    case LoadType.Font: return DefinitionType.Font
  }
}
