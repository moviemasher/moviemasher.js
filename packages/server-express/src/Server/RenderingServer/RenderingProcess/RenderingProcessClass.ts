import fs from 'fs'
import path from 'path'

import {
  idGenerate, Mash, mashInstance, OutputFactory,
  RenderingOutput, RenderingOutputArgs, Errors, CommandFilters,
  RenderingCommandOutput, RenderingResult, CommandOutputs, OutputType,
  EmptyMethod, CommandDescription, CommandDescriptions, CommandOptions, Extension,
  CommandInput, RenderingDescription, AVType, CommandInputs,
  Defined,
  assertTrue,
} from "@moviemasher/moviemasher.js"
import {
  BasenameRendering, ExtensionCommands, ExtensionLoadedInfo
} from '../../../Setup/Constants'
import { RunningCommandFactory } from "../../../RunningCommand/RunningCommandFactory"
import { RenderingProcess, RenderingProcessArgs, RunResult } from "./RenderingProcess"
import { NodeLoader } from '../../../Utilities/NodeLoader'
import { CommandResult } from '../../../RunningCommand/RunningCommand'
import { probingInfoPromise } from '../../../Command/Probing'
import { renderingCommandOutputs, renderingOutputFile } from '../../../Utilities/Rendering'

export type RenderingProcessConcatFileDuration = [string, number]

export class RenderingProcessClass implements RenderingProcess {
  constructor(public args: RenderingProcessArgs) {
    Defined.define(...this.args.definitions)
  }

  concatFile(fileDurations: RenderingProcessConcatFileDuration[]): string {
    const lines = ['ffconcat version 1.0']
    lines.push(...fileDurations.flatMap(fileDuration => {
      const [file, duration] = fileDuration
      return [`file '${file}'`, `duration ${duration}`]
    }))
    return lines.join("\n")
  }

  commandDescriptionMerged(flatDescription: RenderingDescription): CommandDescription {
    const { visibleCommandDescriptions, audibleCommandDescription } = flatDescription
    const descriptions: CommandDescriptions = []
    const length = visibleCommandDescriptions?.length
    if (length) {
      if (length !== 1) throw Errors.internal + 'not flat'

      const [visibleCommandDescription] = visibleCommandDescriptions
      descriptions.push(visibleCommandDescription)
    }
    if (audibleCommandDescription) descriptions.push(audibleCommandDescription)
    if (!descriptions.length) throw Errors.internal + 'descriptions'

    const [description] = descriptions
    const merged = descriptions.length > 1 ? this.commandDescriptionsMerged(descriptions) : description
    return merged
  }
  commandDescriptionsMerged(descriptions: CommandDescriptions): CommandDescription {
    const inputs: CommandInputs = []
    const commandFilters: CommandFilters = []
    const durations: number[] = []
    const types = new Set<AVType>()

    descriptions.forEach(description => {
      const { duration, inputs: descriptionInputs, commandFilters: filters, avType } = description
      types.add(avType)
      if (descriptionInputs) inputs.push(...descriptionInputs)
      if (filters) commandFilters.push(...filters)
      if (duration) durations.push(duration)
    })
    const avType = types.size === 1 ? [...types.values()].pop()! : AVType.Both
    const commandDescription: CommandDescription = { inputs, commandFilters, avType }
    assertTrue(durations.length !== descriptions.length, 'each description has duration')
    commandDescription.duration = Math.max(...durations)
    return commandDescription
  }

  private createDirectoryPromise(directoryPath: string): Promise<void> {
    return fs.promises.mkdir(directoryPath, { recursive: true }).then(EmptyMethod)
  }

  private createFilePromise(filePath: string, content: string): Promise<void> {
    return fs.promises.writeFile(filePath, content).then(EmptyMethod)
  }

  private directoryPromise(): Promise<void> {
    const { outputsPopulated: outputs, args } = this
    const { id } = this
    const { outputDirectory, mash, definitions } = args
    // console.log(this.constructor.name, "directoryPromise", outputDirectory)
    const argsJson = JSON.stringify({ id, outputs, mash, definitions }, null, 2)
    return this.createDirectoryPromise(outputDirectory).then(() => {
      const jsonPath = path.join(outputDirectory, `${BasenameRendering}.json`)
      return this.createFilePromise(jsonPath, argsJson)
    })
  }

  private fileName(commandOutput: RenderingCommandOutput, renderingOutput: RenderingOutput): string {
    const { outputType, videoRate } = commandOutput
    if (outputType !== OutputType.ImageSequence) return renderingOutputFile(commandOutput)
    if (!videoRate) throw Errors.internal + 'videoRate'

    const { format, extension, basename } = commandOutput
    const base = basename || ''
    const ext = extension || format
    const { duration } = renderingOutput
    const framesMax = Math.floor(videoRate * duration) - 2
    const begin = 1
    const lastFrame = begin + (framesMax - begin)
    const padding = String(lastFrame).length
    return `${base}%0${padding}d.${ext}`
  }

  private _id?: string
  get id(): string {
    if (this._id) return this._id

    return this._id = this.args.id || idGenerate()
  }

  private _mashInstance?: Mash
  get mashInstance(): Mash {
    if (this._mashInstance) return this._mashInstance

    const { args } = this
    const {
      mash, cacheDirectory, validDirectories, defaultDirectory, filePrefix
    } = args

    const mashOptions = {
      ...mash,
      preloader: new NodeLoader(cacheDirectory, filePrefix, defaultDirectory, validDirectories)
    }
    return this._mashInstance = mashInstance(mashOptions)
  }

  private outputInstance(commandOutput: RenderingCommandOutput): RenderingOutput {
    const { outputType } = commandOutput
    const { cacheDirectory } = this.args
    // console.log(this.constructor.name, "outputInstance", cacheDirectory)

    const { mashInstance } = this
    const args: RenderingOutputArgs = {
      commandOutput, cacheDirectory, mash: mashInstance
    }
    return OutputFactory[outputType](args)
  }

  private _outputsPopulated?: CommandOutputs
  private get outputsPopulated(): CommandOutputs {
    return this._outputsPopulated ||= renderingCommandOutputs(this.args.outputs)
  }

  renderResultPromise(destinationPath: string, cmdPath: string, infoPath: string, commandOutput: RenderingCommandOutput, commandDescription: CommandDescription): Promise<RenderingResult> {
    const { outputType, avType } = commandOutput
    const { duration, inputs } = commandDescription

    const commandOptions: CommandOptions = {
      output: commandOutput, ...commandDescription
    }

    const options = commandOutput.options!
    switch (outputType) {
      case OutputType.Image:
      case OutputType.Waveform: {
        options['frames:v'] = 1
        break
      }
      default: {
        if (duration) options.t = duration
      }
    }
    if (avType === AVType.Audio) {
      delete commandOutput.videoCodec
      delete commandOutput.videoRate
    } else if (avType === AVType.Video) {
      delete commandOutput.audioCodec
      delete commandOutput.audioBitrate
      delete commandOutput.audioChannels
      delete commandOutput.audioRate
    }

    const command = RunningCommandFactory.instance(this.id, commandOptions)
    const commands = ['ffmpeg', ...command.command._getArguments(), destinationPath]
    const commandsText = commands.join(' ')
    return fs.promises.writeFile(cmdPath, commandsText).then(() => {
      return command.runPromise(destinationPath).then((commandResult: CommandResult) => {
        // console.log(this.constructor.name, "renderResultPromise runPromise", destinationPath, commandResult)
        const renderingResult: RenderingResult = {
          ...commandResult, destination: destinationPath, outputType
        }
        const { error } = commandResult
        if (error) {
          return fs.promises.writeFile(infoPath, JSON.stringify(renderingResult)).then(() => renderingResult)
        }
        return probingInfoPromise(destinationPath, infoPath).then(() => renderingResult)
      })
    })
  }

  private renderingDescriptionPromise(renderingDescription: RenderingDescription): Promise<RenderingDescription> {
    const { visibleCommandDescriptions, commandOutput, audibleCommandDescription } = renderingDescription
    const length = visibleCommandDescriptions?.length
    if (!length || length === 1) {
      // console.log(this.constructor.name, "renderingDescriptionPromise resolved", length)
      return Promise.resolve(renderingDescription)
    }

    const extension = Extension.Mpg
    const {
      options: commandOutputOptions = {},
      audioBitrate, audioChannels, audioCodec, audioRate, outputType,
      ...rest
    } = commandOutput
    const options = { ...commandOutputOptions, an: '', qp: 0 }
    const output: RenderingCommandOutput = {
      ...rest, options, extension, outputType: OutputType.Video
    }
    const { outputDirectory } = this.args
    const temporaryDirectoryName = renderingOutputFile(commandOutput, 'concat')

    const temporaryDirectory = path.join(outputDirectory, temporaryDirectoryName)
    let promise: Promise<void> = this.createDirectoryPromise(temporaryDirectory)
    const fileDurations = visibleCommandDescriptions.map((description, index) => {
      const baseName = `concat-${index}`
      const fileName = `${baseName}.${extension}`
      const destinationPath = path.join(temporaryDirectory, fileName)
      const cmdPath = path.join(temporaryDirectory, `${baseName}.${ExtensionCommands}`)
      const infoPath = path.join(temporaryDirectory, `${baseName}.${ExtensionLoadedInfo}`)
      const { duration } = description
      if (!duration) throw Errors.invalid.duration

      promise = promise.then(() => {
        const renderPromise = this.renderResultPromise(destinationPath, cmdPath, infoPath, output, description).then(EmptyMethod)
        return renderPromise
      })
      const concatFileDuration: RenderingProcessConcatFileDuration = [fileName, duration]
      return concatFileDuration
    })

    const concatFile = this.concatFile(fileDurations)
    const concatFilePath = path.join(temporaryDirectory, 'concat.txt')
    promise = promise.then(() => this.createFilePromise(concatFilePath, concatFile))

    return promise.then(() => {
      const commandInput: CommandInput = { source: concatFilePath }
      const durations = fileDurations.map(([_, duration]) => duration)
      const duration = durations.reduce((total, duration) => total + duration, 0)
      const description: CommandDescription = { inputs: [commandInput], duration, avType: AVType.Video }
      const renderingDescription: RenderingDescription = {
        commandOutput, audibleCommandDescription,
        visibleCommandDescriptions: [description],
      }
      return renderingDescription
    })
  }

  runPromise(): Promise<RunResult> {
    const results: RenderingResult[] = []
    const runResult: RunResult = { results }

    let promise = this.directoryPromise()
    const { outputsPopulated } = this
    outputsPopulated.forEach(output => {
      const { outputType } = output
      const expectDuration = outputType !== OutputType.Image
      // console.log(this.constructor.name, "runPromise", renderingCommandOutput, commandOutput)
      const renderingOutput = this.outputInstance(output)
      // console.log("commandOutput", commandOutput)

      promise = promise.then(() => {
        const outputPromise = renderingOutput.renderingDescriptionPromise(results)
        const flatPromise = outputPromise.then(renderingDescription => {
          return this.renderingDescriptionPromise(renderingDescription)
        })
        return flatPromise.then(flatDescription => {
          const commandDescription = this.commandDescriptionMerged(flatDescription)

          if (expectDuration) {
            const { duration } = commandDescription
            if (!duration) throw Errors.invalid.duration
          }

          const { outputDirectory } = this.args
          const destinationFileName = this.fileName(output, renderingOutput)
          const destPath = `${outputDirectory}/${destinationFileName}`
          const cmdFilename = renderingOutputFile(output, ExtensionCommands)
          const cmdPath = path.join(outputDirectory, cmdFilename)
          const infoFilename = renderingOutputFile(output, ExtensionLoadedInfo)
          const infoPath = path.join(outputDirectory, infoFilename)
          // console.log(this.constructor.name, "runPromise", infoPath)
          const renderPromise = this.renderResultPromise(
            destPath, cmdPath, infoPath, output, commandDescription
          )
          return renderPromise.then(renderingResult => { results.push(renderingResult) })
        })
      })
    })
    return promise.then(() => runResult)
  }
}
