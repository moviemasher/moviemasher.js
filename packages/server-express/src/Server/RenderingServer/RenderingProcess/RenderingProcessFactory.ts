import { RenderingProcessArgs } from "./RenderingProcess"
import { RenderingProcessClass } from "./RenderingProcessClass"

export const renderingProcessInstance = (options: RenderingProcessArgs) => {
  return new RenderingProcessClass(options)
}
