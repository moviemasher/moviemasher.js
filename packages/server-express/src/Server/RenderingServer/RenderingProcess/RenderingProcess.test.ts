import fs from 'fs'
import {
  RenderingCommandOutput, OutputType, LoadType, RenderingInput, OutputFormat,
  ExtPng, CommandOutputs, outputDefaultPopulate
} from "@moviemasher/moviemasher.js"

import { RenderingProcessArgs } from "./RenderingProcess"
import { renderingProcessInstance } from "./RenderingProcessFactory"
import { expectOutputFile } from "../../../../../../dev/test/Utilities/expectOutputFile"
import { renderingProcessTestArgs
} from "../../../../../../dev/test/Utilities/renderingProcessArgs"
import { TestRenderingsDir } from '../../../../../../dev/test/Setup/Constants'
import { renderingInputFromRaw } from '../../../Utilities/Rendering'
import { expandFile } from '../../../Utilities/Expand'

export function renderingProcessArgs(id?: string): RenderingProcessArgs {
  return {
    ...renderingProcessTestArgs(id), mash: {}, outputs: []
  }
}

describe("RenderingProcess", () => {
  const imageUrl = '../shared/image/globe.jpg'
  const videoUrl = '../shared/video.mp4'

  const baseOutput: RenderingCommandOutput = {
    outputType: OutputType.Image,
    format: OutputFormat.Png,
    extension: ExtPng,
  }
  describe("renders", () => {
    const renderingFiles = fs.readdirSync(TestRenderingsDir).filter(p => p.endsWith('.json'))
    test.each(renderingFiles)('%s', async (fileName) => {
      const json = expandFile(`${TestRenderingsDir}/${fileName}`)
      const args = JSON.parse(json) as RenderingProcessArgs
      const { id, outputs } = args
      const renderingArgs = { ...args, ...renderingProcessTestArgs(id) }
      const renderingProcess = renderingProcessInstance(renderingArgs)
      const runPromise = renderingProcess.runPromise()
      const checkPromise = runPromise.then(runResult => {
        const { results } = runResult
        const promises = outputs.map((output, index) => {
          const result = results[index]
          expect(result).toBeDefined()
          const { destination, error } = result
          expect(error).toBeFalsy()
          expect(destination).toBeDefined()

          return expectOutputFile(destination)
        })
        return Promise.all(promises)
      })
      await checkPromise
    })
  })
  describe.skip("render", () => {
    test("renders audio from trimmed video", async () => {
      const output = outputDefaultPopulate({ outputType: OutputType.Audio })
      const outputs: CommandOutputs = [output]
      const renderingInput: RenderingInput = renderingInputFromRaw(LoadType.Video, videoUrl, { trim: 10 })
      const processArgs: RenderingProcessArgs = {
        ...renderingProcessArgs('audio-from-trimmed-video'), outputs, ...renderingInput,
      }
      const renderingProcess = renderingProcessInstance(processArgs)
      const result = await renderingProcess.runPromise()

      await expectOutputFile(result.results[0].destination)
    })
  })

  describe.skip("runPromise", () => {
    test("renders image from image with cover", async () => {
      const output = outputDefaultPopulate({ ...baseOutput, cover: true })
      const outputs: CommandOutputs = [output]
      const renderingInput: RenderingInput = renderingInputFromRaw(LoadType.Image, imageUrl)
      const processArgs: RenderingProcessArgs = {
        ...renderingProcessArgs('image-from-image-square-with-cover'), outputs, ...renderingInput,
      }
      console.log(processArgs)
      const renderingProcess = renderingProcessInstance(processArgs)
      const result = await renderingProcess.runPromise()
      await expectOutputFile(result.results[0].destination)
    })

    test("renders video from image", async () => {
      const output = outputDefaultPopulate({ outputType: OutputType.Video, cover: false })
      const outputs: CommandOutputs = [output]
      const renderingInput: RenderingInput = renderingInputFromRaw(LoadType.Image, imageUrl)
      const processArgs: RenderingProcessArgs = {
        ...renderingProcessArgs('video-from-image-square'), outputs, ...renderingInput,
      }
      const process = renderingProcessInstance(processArgs)
      const result = await process.runPromise()
      await expectOutputFile(result.results[0].destination)
    })

    test("renders image from trimmed video", async () => {
      const output = outputDefaultPopulate({ ...baseOutput })
      const outputs: CommandOutputs = [output]
      const renderingInput: RenderingInput = renderingInputFromRaw(LoadType.Video, videoUrl, { trim: 10 })
      const processArgs: RenderingProcessArgs = {
        ...renderingProcessArgs('image-from-trimmed-video'), outputs, ...renderingInput,
      }
      const renderingProcess = renderingProcessInstance(processArgs)
      const result = await renderingProcess.runPromise()

      await expectOutputFile(result.results[0].destination)
    })

    test("renders video from video with cover", async () => {
      const output = outputDefaultPopulate({ outputType: OutputType.Video, cover: true })
      const outputs: CommandOutputs = [output]
      const renderingInput: RenderingInput = renderingInputFromRaw(LoadType.Video, videoUrl)
      const processArgs: RenderingProcessArgs = {
        ...renderingProcessArgs('video-from-video-with-cover'), outputs, ...renderingInput,
      }
      const renderingProcess = renderingProcessInstance(processArgs)
      const result = await renderingProcess.runPromise()
      const { results } = result
      expect(results.length).toBeGreaterThan(0)
      const [first] = results
      expect(first.error).toBeUndefined()
      await expectOutputFile(first.destination)
    })

    test("renders image sequence from video with cover", async () => {
      const output = outputDefaultPopulate({ ...baseOutput, outputType: OutputType.ImageSequence, cover: true })
      // console.log('output', output)
      const outputs: CommandOutputs = [output]
      const renderingInput: RenderingInput = renderingInputFromRaw(LoadType.Video, videoUrl)
      const processArgs: RenderingProcessArgs = {
        ...renderingProcessArgs('image-sequence-from-video-with-cover'), outputs, ...renderingInput,
      }
      const renderingProcess = renderingProcessInstance(processArgs)
      const result = await renderingProcess.runPromise()
      await expectOutputFile(result.results[0].destination)
    })

    // BUG: test renders as transparent image instead of properly cropped
    test.skip("renders image from image", async () => {
      // expect('image output does not currently support negative values for overlay.y').toBeFalsy()

      const output = outputDefaultPopulate({ ...baseOutput, cover: false })
      const outputs: CommandOutputs = [output]
      const renderingInput: RenderingInput = renderingInputFromRaw(LoadType.Image, imageUrl)
      const processArgs: RenderingProcessArgs = {
        ...renderingProcessArgs('image-from-image-square'), outputs, ...renderingInput,
      }
      const process = renderingProcessInstance(processArgs)
      const result = await process.runPromise()
      await expectOutputFile(result.results[0].destination)
    })
  })
})
