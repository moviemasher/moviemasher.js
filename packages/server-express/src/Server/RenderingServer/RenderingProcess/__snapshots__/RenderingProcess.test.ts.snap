// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RenderingProcess render renders audio from trimmed video 1`] = `
"SHA256=e24dfa97a2bdfa1660de8084f54be754edafcfb85cd0e61533b5e4d19840a6ec
"
`;

exports[`RenderingProcess renders video-from-color.json 1`] = `
"SHA256=566d7ff8add67fdb4cebfb8b6ba73450d32b85ad333d6ce3ed1e459a337d1048
"
`;

exports[`RenderingProcess renders video-from-colors-fade.json 1`] = `
"SHA256=b407bec05d718b8dfb36409f213d72cd43a28dc858a878c8dc6e4b3d2f827fb2
"
`;

exports[`RenderingProcess renders video-from-image.json 1`] = `
"SHA256=7ae2f870b455577af5bbe11232789b8095d4e3c6fba2540aab3508787d486ab5
"
`;

exports[`RenderingProcess renders video-from-image-pan.json 1`] = `
"SHA256=cf17d76075c487acdd34e78c17aa64c553005349990eefab41cfae98f5da8b94
"
`;

exports[`RenderingProcess renders video-from-text-fade.json 1`] = `
"SHA256=0db64fa6f2961f4b0792fbee16cbbf180986a2350004685c189b4a2bd4fd6c2b
"
`;

exports[`RenderingProcess runPromise renders image from image 1`] = `
"SHA256=63d0c0066eb6eaf4149130a1c42e5f13c79532649c43dea2e0c21e48df21b847
"
`;

exports[`RenderingProcess runPromise renders image from image with cover 1`] = `
"SHA256=56c8c4481c7e7492e2d91e0e1553fcdae0723c7ecccddd690fc709bfb57c2ade
"
`;

exports[`RenderingProcess runPromise renders image from trimmed video 1`] = `
"SHA256=2186665f9969418b03710535f624f784d6df5c85cd148c0f4aac24e7e91b5078
"
`;

exports[`RenderingProcess runPromise renders image sequence from video with cover 1`] = `
"SHA256=f50e54fac1449e8b7ae5ab5761784335b0a3a45dc784071e13f2b903e84a9f18
"
`;

exports[`RenderingProcess runPromise renders video from image 1`] = `
"SHA256=9b863fe82883cf0fcd75a7047d567751cbb34cb70093031475b9273f35306665
"
`;

exports[`RenderingProcess runPromise renders video from video with cover 1`] = `
"SHA256=196f7c1dac4b9197939d6edb11fdfa14401ae2ee4e3c7f743bd8f1c67086b3d2
"
`;
